lib/srv/discovery/status.go

```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
 
 	s.updateDiscoveryConfigStatus(result.DiscoveryConfig)
 
+	s.awsEC2Tasks.addFailedEnrollment(
+		awsEC2FailedEnrollmentGroup{
+			integration: result.IntegrationName,
+			issueType:   result.IssueType,
+			accountID:   result.SSMRunEvent.AccountID,
+			region:      result.SSMRunEvent.Region,
+		},
+		&usertasksv1.DiscoverEC2Instance{
+			// TODO(marco): add instance name
+			InvocationUrl:   result.SSMRunEvent.InvocationURL,
+			DiscoveryConfig: result.DiscoveryConfig,
+			DiscoveryGroup:  s.DiscoveryGroup,
+			SyncTime:        timestamppb.New(result.SSMRunEvent.Time),
+			InstanceId:      result.SSMRunEvent.InstanceID,
+		},
+	)
+
+	return nil
+}
+
+// awsEC2Tasks contains the Discover EC2 User Tasks that must be reported to the user.
+type awsEC2Tasks struct {
+	mu sync.RWMutex
+	// instancesIssues maps the Discover EC2 User Task grouping parts to a set of instances metadata.
+	instancesIssues map[awsEC2FailedEnrollmentGroup]map[string]*usertasksv1.DiscoverEC2Instance
+	// groupPending is used to register which groups were changed in memory but were not yet sent upstream.
+	// When upserting User Tasks, if the group is not present in groupPending,
+	// then the cluster already has the latest version of this particular group.
+	groupPending map[awsEC2FailedEnrollmentGroup]struct{}
+}
+
+// awsEC2FailedEnrollmentGroup identifies a UserTask group.
+type awsEC2FailedEnrollmentGroup struct {
+	integration string
+	issueType   string
+	accountID   string
+	region      string
+}
+
+// iterationStarted clears out any in memory issues that were recorded.
+// This is used when starting a new Auto Discover EC2 watcher iteration.
+func (d *awsEC2Tasks) iterationStarted() {
```

Comment:
More of a nit but I think usually methods like this are called something like "reset" or "clear".
```suggestion
func (d *awsEC2Tasks) reset() {
```